enhancements/project-request-limit.md

```diff
@@ -0,0 +1,392 @@
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  namespace: openshift-project-request-limit-operator
+  name: admission-server
+  annotations:
+    service.alpha.openshift.io/serving-cert-secret-name: server-serving-cert
+spec:
+  selector:
+    projectrequestlimits.operator.project.openshift.io/admission-server: "true"
+  ports:
+  - port: 443
+    targetPort: 8443
+```
+ 
+* We define an `APIService` to register the aggregated API provided by the webhook:
+```
+apiVersion: apiregistration.k8s.io/v1
+kind: APIService
+metadata:
+  name: v1.projectrequestlimits.admission.project.openshift.io
+spec:
+  group: projectrequestlimits.admission.project.openshift.io
+  version: v1
+
+  service:
+    name: admission-server
+    namespace: openshift-project-request-limit-operator
+``` 
+
+* We define a `ValidatingWebhookConfiguration` that will allow other components  (`openshift-apiserver` for one) 
+  to reach the webhook via the registered aggregated API:
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingWebhookConfiguration
+webhooks:
+- clientConfig:
+    service:
+      namespace: default
+      name: kubernetes
+      path: /apis/projectrequestlimits.admission.project.openshift.io/v1/validatingadmissionreviews
+    caBundle: KUBE_CA
+```
+
+`openshift-apiserver` reaches out to the admission webhook via `kubernetes.default.svc` as defined in the `ValidatingWebhookConfiguration`.
+`kube-apiserver` reaches out to the webhook via the `admission-server` service in the `openshift-project-request-limit-operator` 
+namespace. At a high level, the call chain looks as below:
+```
+    Project request -> openshift-apiserver -> kube-apiserver -> ProjectRequestLimit webhook.
+```
+
+If we do a deep dive, the call chain to the admission webhook looks as below:
+```
+Project request -> kube-apiserver -> kube aggregator layer inside the kube-apiserver ->
+   openshift-apiserver -> admission layer in openshift-apiserver -> kube-apiserver -> 
+     kube-aggregator layer inside the kube-apiserver -> ProjectRequestLimit webhook 
+```
+
+The webhook server will have the following topology:
+* The webhook is deployed as a `DaemonSet` server.
+* We restrict the webhook server to the master nodes only. Apply appropriate `nodeSelector` and `tolerations` to the 
+  `PodSpec` so that the `Pods` are scheduled on to the master nodes.
+
+
+#### The Operator
+The operator will allow a cluster admin to:
+* Enable or disable `ProjectRequestLimit` admission webhook on a cluster.
+* Specify `limits` on `Project` create request(s) that the `ProjectRequestLimit` admission webhook can enforce.
+* Manage other lifecycle aspects of the operand.
+
+The operator will define a CRD to expose its API. The cluster admin will interact with the operator via a custom resource:
+* We treat the `ProjectRequestLimit` admission webhook (represented by a `DaemonSet`) as a cluster singleton. That mean
+  the operator needs to manage a single instance of the `DaemonSet`.
```

Comment:
not sure I can follow the use of "one instance" here.

Reply:
for a traditional operator, it is possible to have multiple deployments of the operand (each watching a disjoint set of namespaces potentially). In our case, since this is a webhook, by design the operator should not allow this model. 

This is why the operator will reconcile a single CR - `cluster`